accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -621,6 +614,488 @@ It is a warning to use a null aware operator (`?.`, `?[]`, `?..`, `??`, `??=`, o
+in a provisional state, or it may be final.  A **provisional initialization
+inferred type** is guaranteed to be a subtype of the **final initialization
+inferred type**, but allows for additional adjustments to the type around
+nullability as described below.  A **final initialization inferred type** admits
+no further adjustment, and if the analysis completes without error, will be the
+final inferred type of the variable.  In the description below, we use the term
+**initialization inferred type** without qualification to mean either a
+provisional or final **initialization inferred type**.
+
+Unless otherwise noted, all errors specified below are recommended to be
+reported by implementations as failures of inference for the variable in
+question, possibly with additional explanatory text as warranted.
+
+An **initializing assignment** is an assignment to an **untyped variable** when
+it is *promotable via initialization* as defined in
+the
+[flow analysis specification](https://github.com/dart-lang/language/blob/master/resources/type-system/flow-analysis.md). That
+is, the variable is in the **definitely unassigned** state, and the variable
+would have been promoted if it had been declared with type `Object?` and the
+type of the assigned expression were a type of interest for the variable.  Such
+an assignment on a path shall cause the variable in question to be treated by
+the flow analysis on that path as if it were declared with the assigned
+expression as its initializing expression, up until the next join point in the
+program.  This inferred type is the **provisional initialization inferred type**
+for the variable.  The **initialization inferred type** of the variable is
+always treated as a **type of interest** for the purposes of promotion
+(including its non-nullable version if applicable).
+
+```dart
+void test() {
+  var x; // x is an untyped variable
+  x = 3; // x is given the provisional initialization inferred type of `int`
+  var y; // y is an untyped variable
+  y = null; // y is given the provisional initialization inferred type of `Null`
+}
+```
+
+```dart
+void test() {
+  // Error, a type cannot be inferred for x
+  var x;
+  while(b) {
+    x = 3; // Error: not an initializing assignment, so no type for `x`
+  }
+}
+```
+
+If de-promotion is performed to an **untyped variable** which has a
+**provisional initialization inferred type**, then the **provisional
+initialization inferred type** becomes the **final initialization inferred
+type** for the variable, the variable is de-promoted to that type.
+
+
+```dart
+void test() {
+  var x;
+  x = (3 as num); // x is given the provisional initialization inferred type of `num`
+  if (x is! int) return;
+  while(b) { // x is de-promoted to num, and num is made the final initialization inferred type
+    x = 3;
+  }
+}
+```
+
+On joins with back edges in the flow analysis, **untyped variables** are
```

Comment:
Note: the flow analysis spec calls these joins "conservative joins" rather than "joins with back edges".  The rationale is that we have one circumstance where we do conservative joins even though there is no back edge: when analyzing `try` statements.  E.g., in the code below, `x` is promoted to `int` at both the top and bottom of the `try` block, but there is some point during the `try` block where it's not promoted, so we need to demote it during processing of the `catch`.  That's implemented using "conservative join" even though there's no back edge involved.
```dart
Object x = ...;
if (x is! int) return; // Promotes to int
try {
  x = ...; // De-promotes
  ...; // Might throw exception
  if (x is! int) return;
  ...; // Might throw exception
} catch (_) {
  x.isEven; // Error: x not known to be `int`
}
```
I think in all the cases in this doc where you talk about "joins with back edges", the same logic will apply to these sorts of try/catch scenarios.  So you might consider rephrasing some of the text in this doc to use "conservative join" terminology rather than talking about back edges.